### PR TITLE
User authentication feature

### PR DESCRIPTION
**Pull request by MariaAntoniaMarin**
**New feature:** 
As a blog user I should be able to:
1. Log in and log out.
2. Create, edit, delete my blog posts.
3. Add comments in other blog posts only for logged users.

**Change**
- Bcrypt gem to hashing passwords.
- User model associated with articles and comments.
- User controller and modified article and comment controller.
- Authentication controller to create, new, and destroy actions. Also added helper file for manage sessions of authenticated users.
- Modified application controller with authenticate method.
- New user views and modify other views to give support to the new features.
- New routes for login and nested routes to create, update, destroy articles and comments associated.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
+  include AuthenticationHelper
+
+  before_action :authenticate
+
+  private
+
+  def authenticate
+    unless logged_in?
+      store_location
+      flash[:danger] = "Please log in."
```
Comment (estebancordoba) [inline app/controllers/application_controller.rb:11]: What would this line of code do?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
```
Comment (estebancordoba) [inline app/models/user.rb:2]: What is the functionality of `has_secure_password`?

File: app/views/users/_user.json.jbuilder [hunk 3]
```diff
@@ -0,0 +1,2 @@
+json.extract! user, :id, :username, :created_at, :updated_at
```
Comment (estebancordoba) [inline app/views/users/_user.json.jbuilder:1]: What is this file used for?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (MariaAntoniaMarin) [reply app/controllers/application_controller.rb:11]: This line is to alert with a flash message a user that the login failed or is necessary login. But I don't know why it doesn't work. I will review it in detail.

File: app/models/user.rb (hunk 2 above)
Comment (MariaAntoniaMarin) [reply app/models/user.rb:2]: Here I used Bcrypt.

File: app/views/users/_user.json.jbuilder (hunk 3 above)
Comment (MariaAntoniaMarin) [reply app/views/users/_user.json.jbuilder:1]: When I created a user controller with the command, the same command generated views and test, in the views created some json files and style files.

Comment (estebancordoba) [reply app/views/users/_user.json.jbuilder:1]: In fact it is a template that helps RoR to create JSON structures: https://dev.to/ayushmann/how-to-use-json-jbuilder-in-rails-49ef

**Review by estebancordoba — APPROVED** [review]

**Pull request merged**